Add `Prepared By` Constraints

Gabeblis opened the pull request:
# Committer Notes
## Purpose
This PR introduces three constraints aimed at ensuring FedRAMP compliance for the digital authorization package by verifying the `prepared-by` roles/parties. These constraints ensure that all required elements are present and that necessary information is accurately documented.

## Changes
Constraint: `role-defined-prepared-by`
- Validates that there is a defined role with the ID "prepared-by".

Constraint: `responsible-party-prepared-by`
- Ensures that a responsible party is assigned to the prepared-by role.
 
Constraint: `responsible-party-prepared-by-location-valid`
- Validates that the party assigned to the prepared-by role includes a location-uuid with the required address details.

Tests
- Added invalid test data files for all three constraints.
- Added pass/fail YAML files for the three constraints to verify their functionality.
- Updated ssp-all-VALID.xml to contain the required information so that the constraints fire and pass as expected.

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated.?
- [x] If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

aj-stein-gsa commented on src/validations/constraints/content/ssp-all-VALID.xml:
```diff
@@ -12,6 +12,12 @@
+    <role id="prepared-by">
+      <title>Prepared By</title>
+      <description>
+        <p>The organization that prepared this SSP. If developed in-house, this is the CSP itself.</p>
```
I know this is test data, you might want to be careful with that, as _technically_ it is not the CSP that does the preparing (ironic, yes).

```suggestion
        <p>This organization prepared the SSP.</p>
```

aj-stein-gsa commented on src/validations/constraints/content/ssp-responsible-party-prepared-by-INVALID.xml:
```diff
@@ -0,0 +1,13 @@
+    <role id="NOT-prepared-by">
+      <title>Prepared By</title>
+      <description>
+        <p>The organization that prepared this SSP. If developed in-house, this is the CSP itself.</p>
```
Ditto

```suggestion
        <p>This organization prepared the SSP.</p>
```

Gabeblis replied:
fixed

aj-stein-gsa commented on src/validations/constraints/content/ssp-responsible-party-prepared-by-location-valid-INVALID.xml:
```diff
@@ -0,0 +1,22 @@
+        <addr-line>1234 Some Street</addr-line>
+        <!-- missing city -->
+        <state>ME</state>
+        <postal-code>00000</postal-code>
```
Blocking: do we require country-code only for data centers and not all locations? It affects how I move forward with this review.

aj-stein-gsa replied:
Yes sorry I actually look and I was curious. Maybe we should be careful and avoid that for now.

Gabeblis replied:
I wasn't able to find anything saying that the country is required for a location assembly. I just saw that if it isn't included, then `US` is assumed. Maybe we should look a little more into this, or consider revising our current `data-center-us` constraint as it targets every single location assembly. I can definitely open a PR to fix the constraint so that it only targets data centers.

aj-stein-gsa reviewed (changes requested):
Can you address the updated AC requirements in the issue? Sorry again.

Gabeblis commented:
> Can you address the updated AC requirements in the issue? Sorry again.

Resolved here: [75317f8](https://github.com/GSA/fedramp-automation/pull/870/commits/75317f8af2f93d2983635d4ee4c9ee54e3dced6a). I added the check for an embedded address assembly in the party and I added another `-VALID.xml` test file in order to validate both scenarios.

aj-stein-gsa commented on src/validations/constraints/fedramp-external-constraints.xml:
```diff
@@ -235,6 +239,21 @@
+            <expect id="role-defined-prepared-by" target="metadata" test="exists(role[@id eq 'prepared-by'])" level="ERROR">
+                <formal-name>Role Defined Prepared By</formal-name>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/4-expressing-common-fedramp-template-elements-in-oscal/#prepared-by-csp-self-prepared"/>
+                <message>A FedRAMP SSP MUST have a role defined as "prepared-by". </message>
```
Re [FRR13](https://github.com/GSA/fedramp-automation/blob/develop/src/validations/styleguides/STYLE.md#frr113), it would be nice if this message is a little less jargony.

```suggestion
                <message>A FedRAMP SSP MUST define a role for preparing this document. </message>
```

Gabeblis replied:
Resolved: [4191951](https://github.com/GSA/fedramp-automation/pull/870/commits/41919513f6b6f7a43c28ff5c260a9c8b67def846)

Gabeblis commented on src/validations/constraints/fedramp-external-constraints.xml:
```diff
@@ -235,6 +239,21 @@
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/4-expressing-common-fedramp-template-elements-in-oscal/#prepared-by-csp-self-prepared"/>
+                <message>A FedRAMP SSP MUST have a responsible party that references a role defined as "prepared-by". </message>
+            </expect>
+            <expect id="responsible-party-prepared-by-location-valid" target="." test="$prepared-by-party/address[@type='work'] and $prepared-by-party/address/addr-line and $prepared-by-party/address/city and $prepared-by-party/address/state and $prepared-by-party/address/postal-code or $prepared-by-location/address[@type='work'] and $prepared-by-location/address/addr-line and $prepared-by-locati ...
```
It can be one, or both, but never neither.

Gabeblis replied:
Resolved: [b221a4b](https://github.com/GSA/fedramp-automation/pull/870/commits/b221a4bf6c3582dd249927af8e3322739efc3727) and [e8e3d8d](https://github.com/GSA/fedramp-automation/pull/870/commits/e8e3d8d3ecd9e16b7ed3bcedc5b642c9da4d78ce)

Gabeblis commented on src/validations/constraints/content/ssp-responsible-party-prepared-by-location-valid-VALID-1.xml:
This is the case of an address assembly directly in the party itself. Did I misunderstand what Brian said in #805?

aj-stein-gsa replied:
But why would I have an empty address assembly and that pass the test? This inspired my confusion and interest in the same patter in both similar constructs in https://github.com/GSA/fedramp-automation/pull/869/files#r1832038032. But I guess it doesn't matter much.

Gabeblis commented on src/validations/constraints/fedramp-external-constraints.xml:
```diff
@@ -235,6 +239,21 @@
+            <expect id="responsible-party-prepared-by-location-valid" target="." test="$prepared-by-party/address[@type='work'] and $prepared-by-party/address/addr-line and $prepared-by-party/address/city and $prepared-by-party/address/state and $prepared-by-party/address/postal-code or $prepared-by-location/address[@type='work'] and $prepared-by-location/address/addr-line and $prepared-by-locati ...
+                <formal-name>Responsible Party Prepared By Location Valid</formal-name>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/4-expressing-common-fedramp-template-elements-in-oscal/#prepared-by-csp-self-prepared"/>
+                <message>A FedRAMP SSP MUST have a responsible party defined as "prepared-by" that references a party with a valid location. </message>
```
Resolved: [4191951](https://github.com/GSA/fedramp-automation/pull/870/commits/41919513f6b6f7a43c28ff5c260a9c8b67def846)

Gabeblis commented on src/validations/constraints/fedramp-external-constraints.xml:
```diff
@@ -235,6 +239,21 @@
+            <expect id="responsible-party-prepared-by" target="metadata" test="exists(responsible-party[@role-id eq 'prepared-by'])" level="ERROR">
+                <formal-name>Responsible Party Prepared By</formal-name>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/4-expressing-common-fedramp-template-elements-in-oscal/#prepared-by-csp-self-prepared"/>
+                <message>A FedRAMP SSP MUST have a responsible party that references a role defined as "prepared-by". </message>
```
Resolved: [4191951](https://github.com/GSA/fedramp-automation/pull/870/commits/41919513f6b6f7a43c28ff5c260a9c8b67def846)

Gabeblis commented on src/validations/constraints/content/ssp-responsible-party-prepared-by-location-valid-INVALID.xml:
```diff
@@ -0,0 +1,22 @@
+    </role>
+      <location uuid="27b78960-59ef-4619-82b0-ae20b9c709ac">
+      <title>CSP HQ</title>
+      <address type="work">
+        <addr-line>Suite 0000</addr-line>
+        <addr-line>1234 Some Street</addr-line>
+        <!-- missing city -->
+        <state>ME</state>
+        <postal-code>00000</postal-code>
+      </address>
```
Resolved: [d327260](https://github.com/GSA/fedramp-automation/pull/870/commits/d3272603faa0d3bf7ea5009b7dffa47180e3ccb7)

Rene2mt reviewed (commented):
overall, I'm okay with this PR, but we need to determine if we're going to require a country code for locations / addresses.  Perhaps that can be addressed separately from this PR?

aj-stein-gsa commented:
> overall, I'm okay with this PR, but we need to determine if we're going to require a country code for locations / addresses. Perhaps that can be addressed separately from this PR?

For context, one of the other checks for country code is too aggressive and that should be rolled back anyway, I think. So that is why I have not brought it up in this PR yet.

Gabeblis commented:
@aj-stein-gsa @Rene2mt I just opened #885 to address the issue with the country-code checks being too broad in the `data-center` constraints.

kyhu65867 commented on src/validations/constraints/fedramp-external-constraints.xml:
```diff
@@ -312,6 +316,21 @@
+            <expect id="responsible-party-prepared-by-location-valid" target="." test="($prepared-by-party/address[@type='work'] and $prepared-by-party/address/addr-line and $prepared-by-party/address/city and $prepared-by-party/address/state and $prepared-by-party/address/postal-code) or ($prepared-by-location/address[@type='work'] and $prepared-by-location/address/addr-line and $prepared-by-loc ...
+                <formal-name>Responsible Party Prepared By Location Valid</formal-name>
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/general-concepts/4-expressing-common-fedramp-template-elements-in-oscal/#prepared-by-csp-self-prepared"/>
+                <message>A FedRAMP SSP MUST have a responsible party for preparing the document, and that party MUST define an address.</message>
```
If the party MUST define an address, then why is this just a level warning? Shouldn't it then be a level of ERROR? Otherwise should say "SHOULD" or "RECOMMENDED"

Gabeblis replied:
That's a good catch. I do think the address should be required. In that case I'll make this an error.
Resolved: [7332cb6](https://github.com/GSA/fedramp-automation/pull/870/commits/7332cb6d4c46986179a47f2248bcb6013c96b32a)

Gabeblis commented on src/validations/constraints/content/ssp-all-VALID.xml:
```diff
@@ -12,6 +12,12 @@
+    <role id="prepared-by">
+      <title>Prepared By</title>
+      <description>
+        <p>This organization prepared the SSP.</p>
```
That's fair. It could be an org or it could also be the CSP themselves. I'll reword.
Resolved: [7332cb6](https://github.com/GSA/fedramp-automation/pull/870/commits/7332cb6d4c46986179a47f2248bcb6013c96b32a)
